Library/Homebrew/cli/args.rb

```diff
@@ -50,6 +51,111 @@ def flags_only
+      end
+
+      def formulae
+        require "formula"
+        @formulae ||= (downcased_unique_named - casks).map do |name|
+          if name.include?("/") || File.exist?(name)
+            Formulary.factory(name, spec)
+          else
+            Formulary.find_with_priority(name, spec)
+          end
+        end.uniq(&:name)
+      end
+
+      def resolved_formulae
+        require "formula"
+        @resolved_formulae ||= (downcased_unique_named - casks).map do |name|
+          Formulary.resolve(name, spec: spec(nil))
+        end.uniq(&:name)
+      end
+
+      def casks
+        @casks ||= downcased_unique_named.grep HOMEBREW_CASK_TAP_CASK_REGEX
+      end
+
+      def kegs
+        require "keg"
+        require "formula"
+        require "missing_formula"
+        @kegs ||= downcased_unique_named.map do |name|
+          raise UsageError if name.empty?
+
+          rack = Formulary.to_rack(name.downcase)
+
+          dirs = rack.directory? ? rack.subdirs : []
+
+          if dirs.empty?
+            if (reason = Homebrew::MissingFormula.suggest_command(name, "uninstall"))
+              $stderr.puts reason
+            end
+            raise NoSuchKegError, rack.basename
+          end
+
+          linked_keg_ref = HOMEBREW_LINKED_KEGS/rack.basename
+          opt_prefix = HOMEBREW_PREFIX/"opt/#{rack.basename}"
+
+          begin
+            if opt_prefix.symlink? && opt_prefix.directory?
+              Keg.new(opt_prefix.resolved_path)
+            elsif linked_keg_ref.symlink? && linked_keg_ref.directory?
+              Keg.new(linked_keg_ref.resolved_path)
+            elsif dirs.length == 1
+              Keg.new(dirs.first)
+            else
+              f = if name.include?("/") || File.exist?(name)
+                Formulary.factory(name)
+              else
+                Formulary.from_rack(rack)
+              end
+
+              unless (prefix = f.installed_prefix).directory?
+                raise MultipleVersionsInstalledError, rack.basename
+              end
+
+              Keg.new(prefix)
+            end
+          rescue FormulaUnavailableError
+            raise <<~EOS
+              Multiple kegs installed to #{rack}
+              However we don't know which one you refer to.
+              Please delete (with rm -rf!) all but one and then try again.
+            EOS
+          end
+        end
+      end
+
+      private
+
+      def downcased_unique_named
+        # Only lowercase names, not paths, bottle filenames or URLs
+        arguments = if args_parsed
+          remaining
+        else
+          cmdline_args.reject { |arg| arg.start_with?("-") }
+        end
+        arguments.map do |arg|
+          if arg.include?("/") || arg.end_with?(".tar.gz") || File.exist?(arg)
+            arg
+          else
+            arg.downcase
+          end
+        end.uniq
+      end
+
+      def spec(default = :stable)
+        if (args_parsed && HEAD?) || cmdline_args.include?("--HEAD")
```

Comment:
Should/could this logic live inside `HEAD?`?

Reply:
`HEAD?` is one of the keys in OpenStruct, so I don't think so it possible. 
We could extract this logic out to another private method named `HEAD`

Reply:
Yeh, a private method for this sounds good 👍